# API Assignment

Bhavana Balakrishna Rao \
1008221023

Import required libraries

In [14]:
import requests
import matplotlib.pyplot as plt
from bs4 import BeautifulSoup as bs

import random
import json

## Using APIs to Train Models

In [2]:
YELP_TOKEN = "fQOk7zSUJHx-TG_tk77949d_ZuNWDcG-gMLs9smw7MTEKUtYrO9rBu5tqYEm9DyVkcyMaQRHQ9WDUA8TydAK2qd8d_EDi1m0k4IAKmnJLJZ_4LZAZrWhSl6dooSNYXYx"

In [3]:
r = requests.get("https://api.yelp.com/v3/businesses/search?location=Toronto&limit=50", headers={"Authorization": "Bearer %s" % YELP_TOKEN})

In [9]:
print(r.status_code, r.reason, r.content)

200 OK b'{"businesses": [{"id": "r_BrIgzYcwo1NAuG9dLbpg", "alias": "pai-northern-thai-kitchen-toronto-5", "name": "Pai Northern Thai Kitchen", "image_url": "https://s3-media1.fl.yelpcdn.com/bphoto/cO0LcL_SGxZUYrqVeGPJOQ/o.jpg", "is_closed": false, "url": "https://www.yelp.com/biz/pai-northern-thai-kitchen-toronto-5?adjust_creative=pudKXdl82zaTqeQ_-dIWSw&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=pudKXdl82zaTqeQ_-dIWSw", "review_count": 2975, "categories": [{"alias": "thai", "title": "Thai"}], "rating": 4.5, "coordinates": {"latitude": 43.64784, "longitude": -79.38872}, "transactions": [], "price": "$$", "location": {"address1": "18 Duncan Street", "address2": "", "address3": "", "city": "Toronto", "zip_code": "M5H 3G8", "country": "CA", "state": "ON", "display_address": ["18 Duncan Street", "Toronto, ON M5H 3G8", "Canada"]}, "phone": "+14169014724", "display_phone": "+1 416-901-4724", "distance": 3009.3347434869856}, {"id": "iGEvDk6hsizigmXhDKs2Vg", "alias": 

This requests object has information about 50 businesses in Toronto.

In [ ]:
r.json()

{'businesses': [{'alias': 'pai-northern-thai-kitchen-toronto-5',
   'categories': [{'alias': 'thai', 'title': 'Thai'}],
   'coordinates': {'latitude': 43.64784, 'longitude': -79.38872},
   'display_phone': '+1 416-901-4724',
   'distance': 3009.3347434869856,
   'id': 'r_BrIgzYcwo1NAuG9dLbpg',
   'image_url': 'https://s3-media1.fl.yelpcdn.com/bphoto/qEbOKh9rDmTrmHdaUDDnNg/o.jpg',
   'is_closed': False,
   'location': {'address1': '18 Duncan Street',
    'address2': '',
    'address3': '',
    'city': 'Toronto',
    'country': 'CA',
    'display_address': ['18 Duncan Street', 'Toronto, ON M5H 3G8', 'Canada'],
    'state': 'ON',
    'zip_code': 'M5H 3G8'},
   'name': 'Pai Northern Thai Kitchen',
   'phone': '+14169014724',
   'price': '$$',
   'rating': 4.5,
   'review_count': 2960,
   'transactions': [],
   'url': 'https://www.yelp.com/biz/pai-northern-thai-kitchen-toronto-5?adjust_creative=pudKXdl82zaTqeQ_-dIWSw&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=pudK

### Scraping Reviews

Use beautiful soup to scrape reviews using the review url.

In [32]:
review_labels = []
for business in r.json()['businesses']:
  print(business['name'])
  reviews = requests.get("https://api.yelp.com/v3/businesses/%s/reviews" % business['id'], headers={"Authorization": "Bearer %s" % YELP_TOKEN}).json()
  review_url = reviews['reviews'][0]['url']
  page = requests.get(review_url)
  soup = bs(page.content, 'html.parser')
  allreview = soup.find_all("div", class_ = lambda value: value and value.startswith("review__"))
  for review in allreview:
    review_labels.append((review.find("span", class_ = lambda value: value and value.startswith("raw__")).text.rstrip('.'), reviews['reviews'][0]['rating']))

review_labels

Pai Northern Thai Kitchen
Seven Lives Tacos y Mariscos
KINKA IZAKAYA ORIGINAL
Banh Mi Boys
Richmond Station
Byblos
Bang Bang Ice Cream
Ramen Isshin
Khao San Road
Miku
Blaze Pizza
Yasu
Sansotei Ramen
Jacobs & Co. Steakhouse
Insomnia Restaurant & Lounge
Baretto Caffe
Gusto 101
Canoe
Pearl Diver
Wvrst
Mildred's Temple Kitchen
JaBistro
Juicy Dumpling
Scaramouche Restaurant Pasta Bar & Grill
Katsuya
The Stockyards
Fahrenheit Coffee
Ripley's Aquarium of Canada
Alo Restaurant
Lee Restaurant
Hokkaido Ramen Santouka
Eggspectation
Tinuno
Descendant Detroit Style Pizza
GB Hand-pulled Noodles
Wilbur Mexicana
Fresco's Fish & Chips
CN Tower
Black Camel
Saku Sushi
Bar Isabel
KINKA IZAKAYA ANNEX
Banh Mi Boys
Mystic Muffin
Fishman Lobster Clubhouse Restaurant
Ravi Soups
BarChef
Emma's Country Kitchen
Rasa
Beerbistro


[('Pretty authentic thai food in Toronto! Some dishes are a little too salty (the Tom Yum was too salty for me), but everything was quite delicious. Pai is well known and has gained a lot of probability in recent years and now reservations are pretty difficult to make! Not sure if the food if worth the line-up, but it is delicious. There are just other thai restaurants in the city that are just as delicious with no line',
  4),
 ('A quintessential Thai place in Toronto. Reliable service and food options, with ability to change the spice level and make you truly sweat. Good atmosphere to meet with friends or coworkers but always make a reservation because they are packed all the time!',
  4),
 ("Great food, good decoration, perfect place to have a chill dinner. I'd recommend getting their curry. The flavour is justtttt right",
  4),
 ("An easy 4.5 stars. One of the best Thai restaurants I've been to outside of Thailand.We sampled the Green Curry, Panang Curry, Pad Thai, Moo Ping, Chicke

In [44]:
len(review_labels)

500

A total of 500 reviews for 50 restaurents were scrapped.

Creating a dictionary to save the review and the corresponding sentiment (positive or negative).

In [45]:
review_dump = {}
for (x, y) in review_labels:
  review_dump[x] = 'positive' if y > 3 else 'negative'
review_dump

{'"This a review for Alo at Home"This is my first time trying Alo - we made a reservation back in April but it was cancelled due to covid. The menu on the night we ordered takeout, Saturday October 24th, consisted of hen, sides of your choosing, Alo rolls and salad. For the sides we went with the roasted Brussels sprouts and mashed potatoes.Every dish was cooked and salted perfectly and there was an incredible balance of flavours and textures - it was just unbelievable!All of the dishes were packaged well and the reheating instructions were spot on.Hands down the best takeout food I\'ve ever had in my entire life. I would definitely order Alo at Home again and I highly recommend it to anyone looking to have a special meal at home. Five stars': 'positive',
 "$14 for 2 pieces of fish of your choice, for which I ordered cod (they had haddock, and halibut listed inside; about $5 for 1 piece if sold separately/extra), with fries (about $3 extra to make it a poutine, which I did),and a tiny 

Saving the dict into a json file.

In [46]:
with open('data.json', 'w') as fp:
    json.dump(review_dump, fp)

Feature Extraction

In [47]:
review_features = [(x.split(' '), 'positive' if y > 3 else 'negative') for (x, y) in review_labels]
review_features

[(['Pretty',
   'authentic',
   'thai',
   'food',
   'in',
   'Toronto!',
   'Some',
   'dishes',
   'are',
   'a',
   'little',
   'too',
   'salty',
   '(the',
   'Tom',
   'Yum',
   'was',
   'too',
   'salty',
   'for',
   'me),',
   'but',
   'everything',
   'was',
   'quite',
   'delicious.',
   'Pai',
   'is',
   'well',
   'known',
   'and',
   'has',
   'gained',
   'a',
   'lot',
   'of',
   'probability',
   'in',
   'recent',
   'years',
   'and',
   'now',
   'reservations',
   'are',
   'pretty',
   'difficult',
   'to',
   'make!',
   'Not',
   'sure',
   'if',
   'the',
   'food',
   'if',
   'worth',
   'the',
   'line-up,',
   'but',
   'it',
   'is',
   'delicious.',
   'There',
   'are',
   'just',
   'other',
   'thai',
   'restaurants',
   'in',
   'the',
   'city',
   'that',
   'are',
   'just',
   'as',
   'delicious',
   'with',
   'no',
   'line'],
  'positive'),
 (['A',
   'quintessential',
   'Thai',
   'place',
   'in',
   'Toronto.',
   'Reliable',
   '

Training the model to classify the sentiment of the review.

In [48]:
from nltk.sentiment import SentimentAnalyzer
import nltk.sentiment.util
from nltk.classify import NaiveBayesClassifier

random.shuffle(review_features)
training_docs = review_features[:120]
test_docs = review_features[120:]

print("Training: %d, Testing: %d" % (len(training_docs), len(test_docs)))

sentim_analyzer = SentimentAnalyzer()

Training: 120, Testing: 380


In [49]:
all_words_neg = sentim_analyzer.all_words([nltk.sentiment.util.mark_negation(doc) for doc in training_docs])
all_words_neg

['One',
 'line',
 'to',
 'summarize',
 'my',
 'experience',
 'here',
 '-',
 'humble',
 'brunch',
 'done',
 'just',
 'right,',
 'delicious',
 'and',
 'hearty.I',
 'came',
 'here',
 'twice',
 'this',
 'year,',
 'one',
 'time',
 'with',
 'a',
 'larger',
 'group',
 'and',
 'another',
 'time',
 'only',
 'two',
 'of',
 'us.',
 'Tried',
 'two',
 'different',
 'items',
 'of',
 'them:',
 'first',
 'time',
 'was',
 'the',
 'Egg',
 'Benedict',
 'with',
 'peameal',
 'bacon,',
 'the',
 'second',
 'time',
 'was',
 'a',
 'stack',
 'of',
 'pancakes.Egg',
 'Benedict',
 'was',
 'done',
 'nicely:',
 'peameal',
 'bacon',
 'had',
 'crispy',
 'parts',
 'but',
 'not',
 'dry,_NEG',
 'muffin_NEG',
 'was_NEG',
 'of_NEG',
 'the_NEG',
 'right_NEG',
 'thickness,_NEG',
 'egg_NEG',
 'was_NEG',
 'runny_NEG',
 'and_NEG',
 'sauce_NEG',
 'were_NEG',
 'rich._NEG',
 'Nothing_NEG',
 'was_NEG',
 'missing_NEG',
 'at_NEG',
 'all_NEG',
 ':DFor_NEG',
 'the_NEG',
 'pancake,_NEG',
 'I_NEG',
 'especially_NEG',
 'liked_NEG',
 'that

In [50]:
unigram_feats = sentim_analyzer.unigram_word_feats(all_words_neg, min_freq=4)
sentim_analyzer.add_feat_extractor(nltk.sentiment.util.extract_unigram_feats, unigrams=unigram_feats)

In [51]:
training_set = sentim_analyzer.apply_features(training_docs)
test_set = sentim_analyzer.apply_features(test_docs)

In [52]:
trainer = NaiveBayesClassifier.train
classifier = sentim_analyzer.train(trainer, training_set)
for key,value in sorted(sentim_analyzer.evaluate(test_set).items()):
     print('{0}: {1}'.format(key, value))

Training classifier
Evaluating NaiveBayesClassifier results...
Accuracy: 0.9421052631578948
F-measure [negative]: 0.08333333333333333
F-measure [positive]: 0.9701086956521741
Precision [negative]: 0.14285714285714285
Precision [positive]: 0.9571045576407506
Recall [negative]: 0.058823529411764705
Recall [positive]: 0.9834710743801653


The Naive Bayes classifier achieves an accuracy of 94%.